cs/ccxt/base/Throttler.cs

```diff
@@ -62,27 +80,94 @@ private async Task loop()
             }
             else
             {
-                await Task.Delay((int)((double)this.config["delay"] * 1000));
+                await Task.Delay((int)((double)this.config.Delay * 1000));
                 var current = milliseconds();
                 var elapsed = current - lastTimestamp;
                 lastTimestamp = current;
-                var tokens = (double)this.config["tokens"] + ((double)this.config["refillRate"] * elapsed);
-                this.config["tokens"] = Math.Min(tokens, (int)this.config["capacity"]);
+                var tokens = (double)this.config.Tokens + ((double)this.config.RefillRate * elapsed);
+                this.config.Tokens = Math.Min(tokens, (double)this.config.Capacity);
             }
         }
 
     }
 
-    public async Task<Task> throttle(object cost2)
+    private async Task rollingWindowLoop()
     {
+        while (this.running)
+        {
+            lock (queueLock)
+            {
+                if (this.queue.Count == 0)
+                {
+                    this.running = false;
+                    continue;
+                }
+            }
+            (Task, double) first;
+            lock (queueLock)
+            {
+                first = this.queue.Peek();
+            }
+            var task = first.Item1;
+            var cost = first.Item2;
+            var now = milliseconds();
+            var windowSize = Convert.ToDouble(this.config.WindowSize);
+            timestamps.RemoveAll(t => now - t.timestamp >= windowSize);
+            var totalCost = timestamps.Sum(t => t.cost);
+            if (totalCost + cost <= Convert.ToDouble(this.config.MaxWeight))
+            {
+                timestamps.Add((now, cost));
+                await Task.Delay(0);
```

Comment:
shouldn't we run this line at 124th line instead? (after `if (task...)` block and before `lock`)